[mlir][mesh] adding shard-size control

fschlimb opened the pull request:
- Replacing `#mesh.sharding` attribute with operation `mesh.sharding`
  - extended semantics now allow providing optional `halo_sizes` and `sharded_dims_sizes`
  - internally a sharding is represented as a non-IR class `mesh::MeshSharding`

What previously was
```mlir
%sharded0 = mesh.shard %arg0 <@mesh0, [[0]]> : tensor<4x8xf32>
%sharded1 = mesh.shard %arg1 <@mesh0, [[0]]> annotate_for_users : tensor<16x8xf32>
```
is now
```mlir
%sharding = mesh.sharding @mesh0, [[0]] : !mesh.sharding
%0 = mesh.shard %arg0 to %sharding : tensor<4x8xf32>
%1 = mesh.shard %arg1 to %sharding annotate_for_users : tensor<16x8xf32>
```
and allows additional annotations to control the shard sizes:
```mlir
mesh.mesh @mesh0 (shape = 4)
%sharding0 = mesh.sharding @mesh0, [[0]] halo_sizes = [1, 2] : !mesh.sharding
%0 = mesh.shard %arg0 to %sharding0 : tensor<4x8xf32>
%sharding1 = mesh.sharding @mesh0, [[0]] sharded_dims_sizes = [3, 5, 5, 3] : !mesh.sharding
%1 = mesh.shard %arg1 to %sharding1 annotate_for_users : tensor<16x8xf32>
```
- `mesh.shard` op accepts additional optional attribute `force`, useful for halo updates
- Some initial spmdization support for the new semantics
- Support for `tensor.empty` reacting on `sharded_dims_sizes` and `halo_sizes` in the sharding
- New collective operation `mesh.update_halo` as a spmdized target for shardings with `halo_sizes`

@sogartar @yaochengji

fschlimb committed:
New op mesh.sharding replace attribute.
Adding halo_sizes and shard_dims_sizes to sharding.
First spmdization of halo annotated sharding

fschlimb committed:
fixing existing some sharding tests

fschlimb committed:
fixing use of force in spmdization, adding update_halo tests

fschlimb committed:
fixing tests

fschlimb committed:
improved descriptions; more tests

fschlimb committed:
adding test mesh_shape

fschlimb committed:
clang-format

fschlimb committed:
fixing axes parameter to mesh.update_halo, cleanup

sogartar commented on mlir/lib/Dialect/Mesh/Transforms/Spmdization.cpp:
```diff
@@ -450,53 +496,64 @@ reshardOn1DMesh(ImplicitLocOpBuilder &builder, MeshOp mesh,
                                         sourceShard);
 
-  if (reducedSourceSharding == targetSharding) {
+  if (!force && reducedSourceSharding == targetSharding) {
```
Why do we need `force`?

fschlimb replied:
This simply protects existing reshardings because they currently have no logic implemented to deal with `force`.

fschlimb replied:
We need the `force` for cases that require an `update_halo` (see below). Without `force` the sharding can potentially be elided.

sogartar commented on mlir/lib/Dialect/Tensor/IR/ShardingInterfaceImpl.cpp:
```diff
@@ -0,0 +1,103 @@
+    : public ShardingInterface::ExternalModel<EmptyOpShardingInterface,
+                                              tensor::EmptyOp> {
+  SmallVector<utils::IteratorType> getLoopIteratorTypes(Operation *) const {
+    return {utils::IteratorType::parallel};
```
There should be as many iterators as there are dimensions.

fschlimb replied:
ok

sogartar commented on mlir/test/Dialect/Mesh/spmdization.mlir:
```diff
@@ -151,10 +178,86 @@ func.func @incomplete_sharding(
+  %in1: tensor<32x16xi8>
+  // CHECK-SAME: -> tensor<11x16xi8> {
+) -> tensor<32x16xi8> {
+  // CHECK: %[[RES:.*]] = mesh.update_halo %[[IN1]] on @mesh_1d_4  mesh_axes = [0] halo_sizes = [2, 1] : (tensor<11x16xi8>) -> tensor<11x16xi8>
```
I don't understand why this function does need to do anything else then return the same tensor.
The input and output shardings are the same.

fschlimb replied:
See above. This is to deal with inplace semantics.

fschlimb replied:
and it's using `force`

sogartar reviewed (commented):
Thank you for contributing. I have a few remarks.

fschlimb commented on mlir/include/mlir/Dialect/Mesh/Interfaces/ShardingInterface.td:
```diff
@@ -100,7 +100,7 @@ def ShardingInterface : OpInterface<"ShardingInterface"> {
           to shard the op according to shardingOption.
         }],
-        /*retTy=*/"FailureOr<SmallVector<MeshShardingAttr>>",
+        /*retTy=*/"FailureOr<std::vector<MeshSharding>>",
```
The elements are too big for SmallVector (>256 bytes).

fschlimb replied:
I was betting on pass-by-reference and move-semantics - but sure, let's fix this separately.

fschlimb commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.h:
```diff
@@ -33,6 +35,59 @@ using MeshAxesAttr = DenseI16ArrayAttr;
+  bool operator!=(Value rhs) const;
+  bool operator==(const MeshSharding &rhs) const;
+  bool operator!=(const MeshSharding &rhs) const;
+  bool sameExceptConstraint(const MeshSharding &rhs) const;
+  bool sameConstraint(const MeshSharding &rhs) const;
```
Agree. I couldn't come up with a better name. If you have a suggestion, I'd be happy to change.

sogartar replied:
Since we don't have good naming we can be more specific and refactor later.
`equalSplitAndPartialAxes`
`equalHaloAndShardSizes`

fschlimb replied:
done

fschlimb commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -105,22 +105,221 @@ def Mesh_MeshShapeOp : Mesh_Op<"mesh_shape", [
+    Mesh_Sharding:$result
+  );
+  let assemblyFormat = [{
+    $mesh `,` $split_axes
```
ok, will do.

fschlimb commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -105,22 +105,221 @@ def Mesh_MeshShapeOp : Mesh_Op<"mesh_shape", [
+    op. It has 4 possible values:
+    `generic`: is not an allowed value inside a shard attribute.
+
+    5. [Optional] Sizes of halos to be added for each sharded tensor dimension.
+    `halo_sizes`is provided as a flattened 1d array of i64s, 2 values for each sharded dimension.
+    `halo_sizes` = [1, 2] means that the first sharded dimension gets an additional
+    halo of size 1 at the start of the first dimension and a halo size is 2 at its end.
+    `halo_sizes` = [1, 2, 2, 3] defines halos for the first 2 sharded dimensions
+    e.g. the first sharded dimension gets [1,2] halos and the seconds gets [2,3] halos.
+    `?` indicates dynamic halo sizes.
+    
+    6. [Optional] Sizes of sharded dimensions of each shard.
+    `sharded_dims_sizes`is provided as a flattened 1d array of i64s: for each device of the
+    device-mesh one value for each sharded tensor dimension.
+    Assuming a 3d-tensor of shape 32x32x32 with the first 2 dimensions being sharded,
+    `sharded_dims_sizes` = [16, 8, 16, 24] means that the first device of
+    the device-mesh will get a shard of shape 16x8x32 and the second device will get a
+    shard of shape 16x24x32.
+    `?` indicates dynamic shard dimensions.
```
I guess this could be solved with a custom parser so that it accepts nested lists like in the split-axis attribute but using the dynamic/static syntax. I hoped we can either defer or maybe someone has done something like this before and can help/give a pointer. I don't see how a struct would make this easier - can you elaborate?

sogartar replied:
OK, lets leave it like that.

For example
```
halo_sizes = [[1, 2], [2, 3]] # each pair in an array.
halo_sizes = [[1, 2, 3, 4], [5, 6, 7, 8]] # the left/low and right/up sizes.
```

Another thing is what happens with `sharded_dims_sizes` in a dynamic mesh with unknown number of devices.

fschlimb replied:
Good question. It is thinkable to write a dynamic check at least for some basic cases.
For now, I guess it would be a good idea to disallow this case. What do you think?

fschlimb replied:
Added a check in a symbolverifier.

sogartar replied:
Disallowing it is fine. Doing anything meaningful would probably require a formula that computes the shape for each shard, which would be way too complicated.

fschlimb commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -129,15 +328,28 @@ def Mesh_ShardOp : Mesh_Op<"shard", [
     as an operand in subsequent operations. If not, the sharding applies to the
     operation that defines the tensor value.
 
+    4. `force`: A unit attribute requesting an explicit sharding of the data,
+    therefore not allowing to be optimizied away. This is useful in the presence
+    of halos and inplace semantics.
```
That's a possible use. I was primarily focused on inplace semantics like a mutating `insert_slice` which requires a halo-update after the write. I know this goes beyond the current tensor semantics, but we are looking into arrays which do have mutable semantics. An alternative for this case would be an extension of the `ShardingInterface` to indicate which operand a given operation it mutates.

sogartar replied:
Could you add explanation in the doc with an example?
I still don't understand. It would be desirable to avoid the complexity of another attribute.
Shouldn't `insert_slice` know how to handle its spmdization? It would have requirements that the destination-passing style operand and the result are related and that they need to have the same sharding. If this constraint can not be satisfied then resharding will be inserted during sharding propagation.

fschlimb replied:
This might well be a lack of understanding on my side.

An `insert_slice` op does not have a result. So I am not sure I fully follow your line of thought.

This was not meant for the sharding propagation, I agree that this does not need this attribute. 

It was meant to serve the spmdization: when spmdization converts `shard` ops into communication ops, e.g. when the input/operand of a `shard` op is another `shard`. If the shardings of the `shard` ops are identical, the algorithm will simply do nothing, e.g. no sharding/communication will happen. However, an in-place operation will require a halo-update even if the sharding stays identical all the time. The elimination of the `shard` op would be incorrect.

```
...
%x = ...
%t = ...
%sharding = mesh.sharding halo_sizes = [2,2] ...
%t1 = mesh.shard %t %sharding
%x1 = mesh.shard %x %sharding
insert_slice %t1 %x1
%t2 = mesh.shard %t1  // <- this must remain and convert to a halo update
%t3 = mesh.shard %t2
<use %t3>
...
```

One solution could be to let the propagation pass insert an explicit `update_halo` op followed by a `shard` op. This would be pretty specific to halos (and not a s generic as a forced resharding) but maybe that is good enough for now. It might also make passes more complicated because now sharding propagation will create IR with sharding ops as well as explicit communication.

Not sure which one is better. Other suggestions?

fschlimb replied:
Oh, maybe you are suggesting to handle this in the `spmdize` for `insert_slice`. Not sure if how that could handle a change in halo-size.

fschlimb replied:
This conversation gets hidden by github, so I am mentioning @sogartar just in case...

sogartar replied:
`tensor.insert_slice` has a result. It is also a `Pure` operation. Not using its result when this is the intention is an error. Dead code elimination would remove it. In your example, if we are using the result of `tensor.insert_slice`, canonicalization or some other simplification pass should drop the chained `mesh.shard` ops as they refer to the same sharding.
```
...
%x = ...
%t = ...
%sharding = mesh.sharding halo_sizes = [2,2] ...
%t1 = mesh.shard %t %sharding
%x1 = mesh.shard %x %sharding
%t2 = insert_slice %t1 %x1
%t3 = mesh.shard %t2 %sharding
<use %t3>
...
```

fschlimb replied:
Yes, `tensor.insert_slice` does have a return argument. Handling that is relatively straight forward. 
However, array semantics (like numpy) do have inplace semantics (tensors do not!). I was talking about such an inplace `insert_slice` operation which is guaranteed to reuse the existing allocation. The tensor dialect does not have such a thing because it is value-based.
I hope this will become more obvious with my next PR.

fschlimb commented on mlir/lib/Dialect/Mesh/Interfaces/ShardingInterface.cpp:
```diff
@@ -427,11 +441,12 @@ getShardingAttribute(OpResult result, const ShardingOption &shardingOption,
 
-static FailureOr<MeshShardingAttr>
+static FailureOr<MeshSharding>
 getShardingAttribute(OpOperand &opOperand, const ShardingOption &shardingOption,
```
Yes, forgot to change it. I kept it for during the modifications for an easier search...
Will do.

fschlimb commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -105,22 +105,221 @@ def Mesh_MeshShapeOp : Mesh_Op<"mesh_shape", [
+  let description = [{
+    The device/process id is a linearized id of the device/process in the mesh.
+    This operation might be used during spmdization when the shard shape depends
+    on values if the sharding.
```
fixed.

fschlimb commented on mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:
```diff
@@ -398,10 +478,160 @@ bool MeshShardingAttr::operator==(MeshShardingAttr rhs) const {
+        MeshAxesAttr::get(mesh_.getContext(), axis.asArrayRef());
+  }
+
+  auto do_copy = [&](auto src, auto &dst) {
+    dst.resize(src.size());
+    for (auto [i, v] : llvm::enumerate(src)) {
+      dst[i] = v;
+    }
+  };
```
done.

fschlimb commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -879,4 +1055,29 @@ def Mesh_ShiftOp : Mesh_CollectiveCommunicationOpBase<"shift", [
+  ]> {
+  let summary = "Update halo data.";
+  let description = [{
+    Assume all devices hold a tensors with same-sized halo data.
+    Optionally resize to new halo sizes.
```
Done.

fschlimb committed:
review comments: using llvm::copy, docs, sharding syntax

fschlimb commented:
> Thank you for contributing. I have a few remarks.

Thanks for your detailed review! See my comments/modifications.

fschlimb committed:
shardeddims -> shardedDimsSizes

fschlimb commented on mlir/lib/Dialect/Mesh/IR/MeshOps.cpp:
```diff
@@ -150,51 +155,98 @@ static LogicalResult verifyMeshAxes(Location loc, ArrayRef<MeshAxis> axes,
 static void shardShape(const InShape &inShape, const MeshShape &meshShape,
-                       const SplitAxes &splitAxes, OutShape &outShape) {
+                       const SplitAxes &splitAxes, OutShape &outShape,
+                       ArrayRef<int64_t> shardedDims = {},
```
changed `shardedDims` to `shardedDimsSizes`

fschlimb committed:
split_axis = -> split_axes =

fschlimb committed:
UpdateHaloOp with dynamic and static halo sizes

fschlimb commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -879,4 +1056,39 @@ def Mesh_ShiftOp : Mesh_CollectiveCommunicationOpBase<"shift", [
+  let assemblyFormat = [{
+    $input `on` $mesh
+    (`mesh_axes` `=` $mesh_axes^)?
+    `halo_sizes` `=` $dynamic_halo_sizes
```
Fixed by allowing dynamic and static halos

fschlimb committed:
same*Constraint -> equal*

fschlimb commented:
@AntonLydike @tobiasgrosser

yaochengji commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -105,22 +105,222 @@ def Mesh_MeshShapeOp : Mesh_Op<"mesh_shape", [
+    `sharded_dims_sizes`is provided as a flattened 1d array of i64s: for each device of the
+    device-mesh one value for each sharded tensor dimension.
+    Assuming a 3d-tensor of shape 32x32x32 with the first 2 dimensions being sharded,
+    `sharded_dims_sizes` = [16, 8, 16, 24] means that the first device of
```
How would you address the situation where the dim size is dynamic?

fschlimb replied:
How `sharded_dim_sizes` are defined is a matter of the application. It can be constant or a result of some operation which might or might not depend on the actual shard's/tensor's dim sizes.
```MLIR
%c2 = arith.constant 2 : index
%dyn_size1 = arith.constant 33 : index
%dim2 = tensor.dim $tnsr, 2 : (tensor<?x?x?x?xf64>) -> index
%dyn_size2 = arith.add %dim2, %c2 : index
%sharding = mesh.sharding ... sharded_dim_sizes = [22, %dyn_size1, %dyn_size2, 0] ...
```
They only cannot be totally unknown (which wouldn't make sense anyway).

fschlimb committed:
removing force from sharding op

fschlimb committed:
disallowing sharded_dims_sizes on dynamic meshes

fschlimb commented:
@sogartar You're probably right and we can go without the the `force` attribute. I am pretty sure I can handle all update_halo cases without that if I move the generation of `mesh.update_halo` into the `spmdize` method for inplace operations like `insert_slice`. An extra optimization pass should be able to get rid of extra `update_halo` ops, e.g. when the output of one is the input of the other.
I removed `force` from `mesh.shard` and adapted related code/tests.

I also added a check in a symbolverifier to disallow sharded_dims_sizes on dynamic meshes.

As far as I can tell I addressed all your concerns and suggestions.

sogartar commented on mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td:
```diff
@@ -879,4 +1051,40 @@ def Mesh_ShiftOp : Mesh_CollectiveCommunicationOpBase<"shift", [
+  ]> {
+  let summary = "Update halo data.";
+  let description = [{
+    This operation updates halo regions of shards, e.g. if their sharding
+    specified halos and the actual tensor data might have changed
+    on the remote devices. Changes might be caused by mutating operations
+    and/or if the new halo regions are larger than the existing ones.
+
+    Assumes all devices hold tensors with same-sized halo data as specified
+    by `dynamic/static_halo_sizes`.
+
+    `mesh_axes` specifies the tensor axes along which the halo data is updated.
+    Currently each tensor dim can be sharded along a single mesh axis only.
+
+    Optionally resizes to new halo sizes `target_halo_sizes`.
```
I don't understand what this operation should do exactly. We assume there is an overlap of the halos on different devices. How do we know which device holds the "true" value for some tensor region, so that overlapping neighbors can be updated?

fschlimb replied:
This might need adjustments now that `force` is gone. As long as we know the tensor and its sharding we have all data we need. The current definition does not allow splitting a tensor along more than one mesh dimension (the inherited axis argument is 1d). The doc should read
"`mesh_axes` specifies for each tensor axis along which mesh axis its halo data is updated.
    Currently each tensor dim can be sharded along a single mesh axis only."

I started working on a separate PR which will lower this op to MPI. I will fix this there. Hope that's ok.

sogartar reviewed (approved):
@fschlimb, thank you for your contribution. Could you rebase before merging since this PR has been opened for a while and there may be some conflicts.
The commit/pr title should be prefixed with `[mlir][mesh]`.

fschlimb committed:
clarifying doc for update_halo

rengolin commented:
Reverted along with the fixup. Please build the projects that are failing in CI and run the tests above before creating a new PR.
